token/program-2022/src/instruction.rs

```diff
@@ -409,6 +427,125 @@ pub enum TokenInstruction {
         /// The freeze authority/multisignature of the mint.
         freeze_authority: COption<Pubkey>,
     },
+    /// Initialize a new mint for payment systems.
```

Comment:
🤔 The word "payment" is wearing thin already: stables such as USDC and USDT would very much consider themselves to be payment rails and they do not require a transfer fee.

What I think we really want is a mixin concept, there are various independent/optional features that we want to be able to activate for an Account/Mint in the general Token 22 timeframe:
1. Transfer fee
2. Account creation authority (to be able to build allow/deny lists, and charge an account creation fee beyond rent)
3. Confidential token support.  Currently this is implemented as a PDA, but being able to build it directly into SPL Token would be a significant simplification
4. A joint Mint/Account type for NTFs (remove the need to have 1 Mint + 1 Account for each NTF)

What if the extended Mint/Account data was the base mint/account followed by TLV scheme that contained an arbitrary number of extensions.

Reply:
It's certainly not a perfect name, but I'll take it over `Mint2`.  If you want to go ahead and call it `MixinMint` to start the framework concept, I'm game.

In that case, we might have to nix some of the `TransferFee`-specific instructions in favor of more general ones.  `2. Account creation authority` and `4. joint mint/account` seem pretty simple to tack on, but maybe we should have a chat about `3. Confidential token support` to make sure this design doesn't make it gross later.

Thanks for laying out all of the other features that we're looking for. A TLV scheme is probably our best bet to maintain all of these separate features all in the same account and expand continuously without breaking everyone anytime a new mint type comes out.

It's too bad, I was really hoping that some of those mixin features could get done in separate programs. I'll just hope that a mixin model will make it easier for people to write their own programs in the future :crossed_fingers:

Reply:
> It's certainly not a perfect name, but I'll take it over Mint2.

No argument from me on that one! :)

> If you want to go ahead and call it MixinMint to start the framework concept, I'm game.

Here's an idea: we could just use the existing Account and Mint structs and then append TLV data at the end of each.  No need to define new Account/Mint/Multisig structs at all!

If the account data length is not 82, 165, or 355 (legacy Mint/Account, and Multisig accounts), the account data layout would be:
- 0..(81 or 164): existing Mint/Account struct
- 82..164: zero padding only present if this is a Mint.
- 165: a u8 that indicates the account type (Account = 1, Mint = 2).
- TLV entries...

Depending on the TLV entries present, the account data could still end up being 355 bytes long and be confused with a Multisig account. If this happens then a null TLV entry must also be added to ensure the data length never equals the Multisig account data length

> It's too bad, I was really hoping that some of those mixin features could get done in separate programs.

The ZkToken work is an example of exactly that.  It wasn't too painful honestly, but a couple things did feel quite unnatural:
a. One can close the primary token account and orphan the ZkToken account
b. All the ZkToken instructions need the primary token account to look up the owner (so more accounts to pass in/decode)
c. Frozen account handling logic needed to be duplicated in the ZkToken transfer processor

> but maybe we should have a chat about 3. Confidential token support to make sure this design doesn't make it gross later.

For sure.  I think it's pretty straightforward though:
1. For Mints, the TLV value would be the [ZkMint](https://github.com/solana-labs/spl-zk-token/blob/662d0b51105dd7b9cc6aae60f353bb8676f302ae/program/src/state.rs#L13) struct minus the `mint` field (since that's now implicit)
2. For Accounts, the TLV value would be the [ZkAccount](https://github.com/solana-labs/spl-zk-token/blob/662d0b51105dd7b9cc6aae60f353bb8676f302ae/program/src/state.rs#L58) struct minus the `mint` and `token_account` fields.
and then most of the [ZkTokenInstruction](https://github.com/solana-labs/spl-zk-token/blob/662d0b51105dd7b9cc6aae60f353bb8676f302ae/program/src/instruction.rs#L70)s end up moving into SPL Token as well.


The only major downside I currently see with a TLV approach is that it makes RPC `getProgramAccounts` `memcmp` searches harder (impossible?), for example [this](https://spl.solana.com/token#finding-all-token-accounts-for-a-specific-mint), since the byte offset for a particular TLV tag cannot be relied upon.  Maybe that's ok though?  The legacy Account/Mint fields remain searchable, even for extended accounts/mints

Reply:
> Here's an idea: we could just use the existing Account and Mint structs and then append TLV data at the end of each. No need to define new Account/Mint/Multisig structs at all!

I really really like this idea.  It resolves a lot of my hangups with how clients will use this, and all mixins can just have an unpack which tries to find and deserialize, returning an `Option`. In processors, if an instruction is impacted by a particular mixin, it just needs to extract that bit, perform whatever extra logic is needed, and you're good to go!

This means we won't need quite so many instructions in the end and some initializers that take lists of mixins + any required parameters.  I'll update this PR with some interfaces about how this would look.

> For Mints, the TLV value would be the ZkMint struct minus the mint field (since that's now implicit)
> For Accounts, the TLV value would be the ZkAccount struct minus the mint and token_account fields and then most of the ZkTokenInstructions end up moving into SPL Token as well.

Now that I've read through the program, you're absolutely right, this'll be really straightforward to include as a mixin.  My one question: do you still envision having `Deposit` / `Withdraw` logic to "convert" between zk and non-zk tokens?  It seems like that should be done through an external swapping program.

Reply:
> My one question: do you still envision having Deposit / Withdraw logic to "convert" between zk and non-zk tokens? 

Yep, that's what I was thinking still.

> It seems like that should be done through an external swapping program.

I guess I don't see a benefit to externalizing that as it's still the same token whether it's on the normal or zk side.

Reply:
oh ok, so that means the ZK token mint would support both normal accounts and ZK accounts? and you would `deposit` from your normal account to your ZK account? or would the same account somehow handle both?

I thought it would be two separate tokens, one for just ZK and another for normal, and have a program in between to do swaps. The idea being that ZK tokens are just confidential wrapped tokens around the base ones.

Reply:
Yeah, I was still thinking the same account could handle both.  The zk tokens are confidential, not anonymous, so there doesn’t seem to be much benefit to making users create a separate token account to hold them since normal/zk sides share the same mint, supply, and account owner. 

When the user/wallet creates their token account, they’ll just need to decide if reserving space for the zk mixin is worth it for them (but the ATA program needs to change to support mixins regardless). If they change their mind later, the wallet would need to close and recreate their token account.

Reply:
Gotcha, wfm